simpa/core/device_digital_twins/illumination_geometries/ring_illumination.py

```diff
@@ -0,0 +1,122 @@
+
+
+class RingIlluminationGeometry(IlluminationGeometryBase):
+    """
+    Defines a ring illumination geometry.
+    The device position is defined as the center of the ring.
+
+    Note: To create a ring light which illuminates a square tissue with the same center and any inner radius r,
+    create the following geometry using the tissue width:
+
+    >>> ring_light = RingIlluminationGeometry(inner_radius_in_mm=r, outer_radius_in_mm=tissue_width / 2., device_position_mm=np.array([tissue_width / 2., tissue_width / 2., 0]))
+    """
+
+    def __init__(self,
+                 outer_radius_in_mm: float = 1,
+                 inner_radius_in_mm: float = 0,
+                 lower_angular_bound: float = 0,
+                 upper_angular_bound: float = 0,
+                 device_position_mm: typing.Optional[np.ndarray] = None,
+                 source_direction_vector: typing.Optional[np.ndarray] = None,
+                 field_of_view_extent_mm: typing.Optional[np.ndarray] = None):
+        """
+        :param outer_radius_in_mm: The outer radius of the ring in mm.
+        :param inner_radius_in_mm: The inner radius of the ring in mm. If 0, should match the disk illumination.
+        :param lower_angular_bound: The lower angular bound in radians. If both bounds are 0, than no bound is applied.
+        Note that the bound of 0 starts from the x-axis on the right side and is applied clockwise.
+        :param upper_angular_bound: The upper angular bound in radians. If both bounds are 0, than no bound is applied.
+        Note that the bound is applied clockwise in relation to the lower bound.
+        :param device_position_mm: The device position in mm, the center of the ring.
+        If None, the position is defined as [0, 0, 0].
+        :param source_direction_vector: Direction of the illumination source.
+        If None, the direction is defined as [0, 0, 1].
+        :param field_of_view_extent_mm: Field of view which is defined as a numpy array of the shape \
+        [xs, xe, ys, ye, zs, ze], where x, y, and z denote the coordinate axes and s and e denote the start and end \
+        positions.
+        """
+        if device_position_mm is None:
+            device_position_mm = np.zeros(3)
+
+        if source_direction_vector is None:
+            source_direction_vector = np.array([0, 0, 1])
+
+        super(RingIlluminationGeometry, self).__init__(device_position_mm=device_position_mm,
+                                                       source_direction_vector=source_direction_vector,
+                                                       field_of_view_extent_mm=field_of_view_extent_mm)
+
+        assert inner_radius_in_mm >= 0, inner_radius_in_mm
+        assert outer_radius_in_mm >= inner_radius_in_mm, \
+            f"The outer radius ({outer_radius_in_mm}) has to be at least as large " \
+            f"as the inner radius ({inner_radius_in_mm})!"
+
+        assert lower_angular_bound >= 0, lower_angular_bound
```

Comment:
Please add an assertion error string saying that the lower_angular_bound is negative